File: backend/app/ifc/kategorien.py
"""Fachliche Kategorien der CDE — an der richtigen HOEHE im IFC-Baum, nicht als Aufzaehlung.

Bis 2026-09-11 stand im Server-Verbundlauf zweimal `klasse == "IFCEARTHWORKSCUT"`.
Ein Untertyp oder eine andere Schreibweise waere still durchgefallen. Hier steht
die WURZEL; ob eine Klasse darunter liegt, beantwortet der Schema-Schnappschuss
(`schema.ist_untertyp`).

Spiegel von client/src/features/cde/services/Kategorien.js — der Test haelt die
Wurzeln zeilengleich (Muster test_bezugssysteme.py). Rein: der API-Server
importiert diese Datei, dort gibt es kein ifcopenshell.
"""
from . import schema as S

# Ein Aushub im Sinne des Erdbau-Dokuments: sein Wirt ist das Ur-Gelaende.
AUSHUB_WURZELN = ("IfcEarthworksCut",)


def ist_aushub(klasse) -> bool:
    """Liegt diese Klasse (gleich wie geschrieben) unter einer Aushub-Wurzel?"""
    return any(S.ist_untertyp(klasse, w) for w in AUSHUB_WURZELN)

File: client/src/features/cde/services/Kategorien.js
/**
 * Fachliche Kategorien — an der richtigen HÖHE im IFC-Baum, nicht als Aufzählung.
 *
 * WARUM (2026-09-11): „was ist linear" stand zweimal wortgleich im Code
 * (KgClassifier, GeometryResolver), jedes Mal als Liste mit exaktem
 * Namensvergleich — samt `IFCDUCT`, einer Klasse, die kein Schema kennt, und
 * ohne `IFCCONVEYORSEGMENT` oder die StandardCase-Fassungen. Eine Liste deckt,
 * was jemand beim Schreiben im Kopf hatte; eine Wurzel im Baum deckt jeden
 * Untertyp, auch künftige.
 *
 * Die Frage „liegt X unter Y" beantwortet EINE Stelle: `vererbungskette`
 * (Typprofile.js, aus dem erzeugten Wörterbuch). Hier stehen nur die Wurzeln —
 * fachliche Entscheidungen, keine Schemafakten.
 *
 * Darstellungstabellen (Linienstile, Farben, Ebenen-Icons) bleiben Tabellen
 * mit exakten Schlüsseln: dort IST die einzelne Klasse gemeint.
 * `test/kategorienWaechter.test.js` hält fest, dass keine der alten Listen
 * zurückkommt und dass jeder IFC-Name im Code eine Klasse ist.
 */
import { vererbungskette } from './bauform/Typprofile.js';

/** Achsförmige Bauteile: eine Länge je Stück ist sinnvoll, die Achse lässt sich ableiten. */
export const LINEARE_WURZELN = Object.freeze(['IFCFLOWSEGMENT', 'IFCBEAM', 'IFCMEMBER', 'IFCKERB']);

/** Schächte, wie der Längsschnitt sie als Balken zeichnet. */
export const SCHACHT_WURZELN = Object.freeze(['IFCDISTRIBUTIONCHAMBERELEMENT']);

/**
 * Aushub im Sinne des Erdbau-Dokuments: sein Wirt ist das Ur-Gelände.
 * Spiegel: backend/app/ifc/kategorien.py (`AUSHUB_WURZELN`) — test_kategorien.py
 * hält beide gleich.
 */
export const AUSHUB_WURZELN = Object.freeze(['IFCEARTHWORKSCUT']);

const _ketten = new Map();

/** Liegt die Kategorie (oder ihr Altname) unter einer dieser Wurzeln? Unbekannt heisst nein. */
export function istUnter(kategorie, wurzeln) {
    const k = String(kategorie ?? '').toUpperCase().trim();
    if (!k) return false;
    let kette = _ketten.get(k);
    if (!kette) {
        kette = vererbungskette(k);
        _ketten.set(k, kette);
    }
    return wurzeln.some(w => kette.includes(w));
}

export const istLinear = kategorie => istUnter(kategorie, LINEARE_WURZELN);
export const istSchacht = kategorie => istUnter(kategorie, SCHACHT_WURZELN);
export const istAushub = kategorie => istUnter(kategorie, AUSHUB_WURZELN);
